src/k2/module/k23si/Module.cpp

```diff
@@ -835,90 +918,117 @@ bool K23SIPartitionModule::_makeProjection(dto::SKVRecord::Storage& fullRec, dto
+                K2LOG_D(log::skvsvr, "Partition: {}, succeeded creating TR. Processing write for {}", _partition, request.mtr);
+                request.designateTRH = false;
+                return _processWrite(std::move(request), deadline);
+            });
+    }
+
+    return _processWrite(std::move(request), deadline);
+}
+
+seastar::future<std::tuple<Status, dto::K23SIWriteResponse>>
+K23SIPartitionModule::_processWrite(dto::K23SIWriteRequest&& request, FastDeadline deadline) {
+    K2LOG_D(log::skvsvr, "Partition: {}, processing write: {}", _partition, request);
+    auto& vset = _indexer[request.key];
+    Status validateStatus = _validateWriteRequest(request, vset);
+
+    if (!validateStatus.is2xxOK()) {
+        if (vset.empty()) {
+            // remove the key from indexer if there are no versions for it
+            _indexer.erase(_indexer.find(request.key));
         }
+        // we may come here after a TRH create. Make sure to flush that
+        return _respondAfterFlush(std::move(validateStatus), dto::K23SIWriteResponse{});
     }
 
     // check to see if we should push or is this a write from same txn
-    if (versions.WI.has_value() && versions.WI->txnId.mtr != request.mtr) {
+    if (vset.WI.has_value() && vset.WI->txnId.mtr != request.mtr) {
         // this is a write request finding a WI from a different transaction. Do a push with the remaining
         // deadline time.
         K2LOG_D(log::skvsvr, "Partition: {}, different WI found for key {}", _partition, request.key);
-        return _doPush(request.collectionName, request.key, versions.WI->txnId, request.mtr, deadline)
+        return _doPush(request.collectionName, request.key, vset.WI->txnId, request.mtr, deadline)
             .then([this, request = std::move(request), deadline](auto&& retryChallenger) mutable {
-                if (retryChallenger) {
-                    K2LOG_D(log::skvsvr, "Partition: {}, write push retry for key {}", _partition, request.key);
-                    return handleWrite(std::move(request), deadline);
+                if (!retryChallenger) {
+                    // challenger must fail. Flush in case a TR was created during this call to handle write
+                    K2LOG_D(log::skvsvr, "Partition: {}, write push challenger lost for key {}", _partition, request.key);
+                    return _respondAfterFlush(dto::K23SIStatus::AbortConflict("incumbent txn won in write push"), dto::K23SIWriteResponse{});
                 }
-                // challenger must fail
-                K2LOG_D(log::skvsvr, "Partition: {}, write push challenger lost for key {}", _partition, request.key);
-                return RPCResponse(dto::K23SIStatus::AbortConflict("incumbent txn won in write push"), dto::K23SIWriteResponse{});
+
+                K2LOG_D(log::skvsvr, "Partition: {}, write push retry for key {}", _partition, request.key);
+                return handleWrite(std::move(request), deadline);
```

Comment:
Maybe should call directly _processWrite()?

Reply:
yup